File: agents/Day-06-Learning-Coach/backend/app/api/endpoints/quizzes.py
"""
Quizzes endpoints for the Learning Coach Agent.
"""

from typing import List, Optional, Dict, Any

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.db.base import get_db
from app.core.quiz_generator import QuizGenerator

router = APIRouter()

# Initialize the quiz generator
quiz_generator = QuizGenerator()

class QuizQuestion(BaseModel):
    """Quiz question schema."""
    
    question: str
    options: List[str]
    correct_answer: int
    explanation: str


class Quiz(BaseModel):
    """Quiz schema."""
    
    id: str
    title: str
    description: str
    topic: str
    difficulty: str
    questions: List[QuizQuestion]
    estimated_time_minutes: int
    created_at: str
    user_id: Optional[str] = None


class QuizAttempt(BaseModel):
    """Quiz attempt schema."""
    
    quiz_id: str
    user_id: str
    answers: List[int]
    score: float
    completed: bool
    date: str


@router.get("/", response_model=List[Dict[str, Any]])
def get_quizzes(
    topic: Optional[str] = None,
    difficulty: Optional[str] = None,
    user_id: Optional[str] = None,
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db)
):
    """
    Get all quizzes, optionally filtered by topic, difficulty, and user_id.
    """
    try:
        quizzes = quiz_generator.get_quizzes(
            topic=topic,
            difficulty=difficulty,
            user_id=user_id,
            skip=skip,
            limit=limit
        )
        return quizzes
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error retrieving quizzes: {str(e)}")


@router.get("/{quiz_id}", response_model=Dict[str, Any])
def get_quiz(
    quiz_id: str,
    db: Session = Depends(get_db)
):
    """
    Get a specific quiz by ID.
    """
    try:
        quiz = quiz_generator.get_quiz(quiz_id)
        if not quiz:
            raise HTTPException(status_code=404, detail="Quiz not found")
        return quiz
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error retrieving quiz: {str(e)}")


@router.post("/generate", response_model=Dict[str, Any])
async def generate_quiz(
    topic: str,
    difficulty: str = "beginner",
    num_questions: int = 5,
    question_types: List[str] = ["multiple_choice"],
    learning_objectives: List[str] = [],
    user_id: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """
    Generate a new quiz based on the given parameters.
    """
    try:
        quiz = await quiz_generator.generate_quiz(
            topic=topic,
            difficulty=difficulty,
            num_questions=num_questions,
            question_types=question_types,
            learning_objectives=learning_objectives,
            user_id=user_id
        )
        return quiz
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error generating quiz: {str(e)}")


@router.post("/{quiz_id}/submit", response_model=Dict[str, Any])
def submit_quiz_attempt(
    quiz_id: str,
    answers: List[int],
    user_id: str,
    db: Session = Depends(get_db)
):
    """
    Submit a quiz attempt and get the results.
    """
    try:
        result = quiz_generator.submit_quiz_attempt(
            quiz_id=quiz_id,
            user_id=user_id,
            answers=answers
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error submitting quiz attempt: {str(e)}")


@router.get("/{quiz_id}/attempts", response_model=List[QuizAttempt])
def get_quiz_attempts(
    quiz_id: str,
    user_id: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """
    Get all attempts for a specific quiz, optionally filtered by user_id.
    """
    try:
        attempts = quiz_generator.get_quiz_attempts(
            quiz_id=quiz_id,
            user_id=user_id
        )
        return attempts
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error retrieving quiz attempts: {str(e)}")


@router.post("/recommend", response_model=Dict[str, Any])
async def recommend_quiz(
    topic: str,
    user_id: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """
    Recommend a quiz based on the user's learning history and the given topic.
    """
    try:
        quiz = await quiz_generator.recommend_quiz(
            topic=topic,
            user_id=user_id
        )
        return quiz
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error recommending quiz: {str(e)}")

File: agents/Day-06-Learning-Coach/frontend/src/api/quizzes.ts
import api from './api';

export interface QuizQuestion {
  question: string;
  options: string[];
  correct_answer: number;
  explanation: string;
  question_type: string;
}

export interface Quiz {
  id: string;
  title: string;
  description: string;
  topic: string;
  difficulty: string;
  questions: QuizQuestion[];
  estimated_time_minutes: number;
  created_at: string;
  learning_objectives?: string[];
  tags?: string[];
  user_id?: number;
}

export interface QuizAttempt {
  id: string;
  quiz_id: string;
  user_id?: number;
  answers: number[];
  score: number;
  correct_answers: number;
  total_questions: number;
  feedback?: any;
  completed_at: string;
}

export interface GenerateQuizRequest {
  topic: string;
  difficulty?: string;
  num_questions?: number;
  question_types?: string[];
  learning_objectives?: string[];
  user_id?: number;
}

export const getQuizzes = async (
  topic?: string,
  difficulty?: string,
  user_id?: number
): Promise<Quiz[]> => {
  let url = '/quizzes';
  const params = new URLSearchParams();
  
  if (topic) params.append('topic', topic);
  if (difficulty) params.append('difficulty', difficulty);
  if (user_id) params.append('user_id', user_id.toString());
  
  if (params.toString()) {
    url += `?${params.toString()}`;
  }
  
  const response = await api.get<Quiz[]>(url);
  return response.data;
};

export const getQuiz = async (id: string): Promise<Quiz> => {
  const response = await api.get<Quiz>(`/quizzes/${id}`);
  return response.data;
};

export const generateQuiz = async (request: GenerateQuizRequest): Promise<Quiz> => {
  const response = await api.post<Quiz>('/quizzes/generate', request);
  return response.data;
};

export const submitQuizAttempt = async (
  quiz_id: string,
  answers: number[],
  user_id?: number
): Promise<QuizAttempt> => {
  const response = await api.post<QuizAttempt>(`/quizzes/${quiz_id}/submit`, {
    answers,
    user_id: user_id?.toString()
  });
  return response.data;
};

// Function to save a quiz from the chat interface
export const saveQuizFromChat = async (quiz: any): Promise<Quiz> => {
  // The quiz from chat should already be in the correct format
  // We just need to make sure it has the required fields
  const generateRequest: GenerateQuizRequest = {
    topic: quiz.topic,
    difficulty: quiz.difficulty,
    num_questions: quiz.questions.length,
    question_types: ['multiple_choice'],
    learning_objectives: quiz.learning_objectives || []
  };
  
  // We're not actually generating a new quiz, but using the API endpoint
  // to save the existing quiz to the database
  const response = await api.post<Quiz>('/quizzes/generate', {
    ...generateRequest,
    // Include the existing quiz data
    existing_quiz: quiz
  });
  
  return response.data;
};
